connectorx/src/sources/mysql/mod.rs

```diff
@@ -0,0 +1,284 @@
+    }
+
+    fn fetch_metadata(&mut self) -> Result<()> {
+        assert!(self.queries.len() != 0);
+
+        let mut conn = self.pool.get()?;
+        let mut success = false;
+        let mut error = None;
+        for query in &self.queries {
+            // assuming all the partition queries yield same schema
+            match conn.query_first(&limit1_query(query)?[..]) {
+                Ok(res) => {
+                    let row: Row = res.unwrap();
+                    self.names = row
+                        .columns_ref()
+                        .into_iter()
+                        .map(|col| col.name_str().to_string())
+                        .collect();
+                    self.schema = row
+                        .columns_ref()
+                        .into_iter()
+                        .map(|col| MysqlTypeSystem::from(&col.column_type()))
+                        .collect();
+                    success = true;
+                    break;
+                }
+                Err(e) => {
+                    debug!("cannot get metadata for '{}', try next query: {}", query, e);
+                    error = Some(e);
+                }
+            }
+        }
+
+        if !success {
+            throw!(anyhow!(
+                "Cannot get metadata for the queries, last error: {:?}",
+                error
+            ))
+        }
+
+        Ok(())
+    }
+
+    fn names(&self) -> Vec<String> {
+        self.names.clone()
+    }
+
+    fn schema(&self) -> Vec<Self::TypeSystem> {
+        self.schema.clone()
+    }
+
+    fn partition(self) -> Result<Vec<Self::Partition>> {
+        let mut ret = vec![];
+        for query in self.queries {
+            let conn = self.pool.get()?;
+            ret.push(MysqlSourcePartition::new(
+                conn,
+                &query,
+                &self.schema,
+                self.buf_size,
+            ));
+        }
+        Ok(ret)
+    }
+}
+
+pub struct MysqlSourcePartition {
+    conn: MysqlConn,
+    query: String,
+    schema: Vec<MysqlTypeSystem>,
+    nrows: usize,
+    ncols: usize,
+    buf_size: usize,
+    // _protocol: PhantomData<P>,
+}
+
+impl MysqlSourcePartition {
+    pub fn new(conn: MysqlConn, query: &str, schema: &[MysqlTypeSystem], buf_size: usize) -> Self {
+        Self {
+            conn,
+            query: query.to_string(),
+            schema: schema.to_vec(),
+            nrows: 0,
+            ncols: schema.len(),
+            buf_size,
+            // _protocol: PhantomData,
+        }
+    }
+}
+
+impl SourcePartition for MysqlSourcePartition {
+    type TypeSystem = MysqlTypeSystem;
+    type Parser<'a> = MysqlSourcePartitionParser<'a>;
+
+    fn prepare(&mut self) -> Result<()> {
+        self.nrows = match get_limit(&self.query)? {
+            // now get_limit using PostgreDialect
+            None => {
+                let row: Option<usize> = self.conn.query_first(&count_query(&self.query)?).unwrap();
+                row.unwrap()
```

Comment:
Would be better if we can try to throw an error instead of using unwrap. Same for rest of the code.

Reply:
`    fn prepare(&mut self) -> Result<()> {
        self.nrows = match get_limit(&self.query)? {
            // now get_limit using PostgreDialect
            None => {
                let row: Option<usize> = self.conn.query_first(&count_query(&self.query)?).unwrap();
                row.unwrap_or_else(|| panic!("failed to get the count of query: {:?}", self.query))
            }
            Some(n) => n,
        };
        Ok(())
    }`
I changed the code like this, I am not sure if it is suitable or not.

Reply:
fn prepare(&mut self) -> Result<()> {
        self.nrows = match get_limit(&self.query)? {
            // now get_limit using PostgreDialect
            None => {
                let row: Option<usize> = self.conn.query_first(&count_query(&self.query)?).unwrap();
                row.unwrap_or_else(|| panic!("failed to get the count of query: {:?}", self.query))
            }
            Some(n) => n,
        };
        Ok(())
    }

I changed the code like this, using `unwrap_or_else`. Does that make sense or not?